make-zip

```diff
@@ -1,80 +1,183 @@
-        with open(os.path.join(tmpd, "wifi.yaml"), "w") as f:
-            yaml.safe_dump( { "ssid": "robot-{}".format(args.team_tla),
-                              "key": y[args.team_tla] }, f,
-                            default_flow_style=False )
-            default_wifi = False
-
-if default_wifi:
-    print("Default wifi keys used")
-
-# Temporary directory for the zipfile to reside in
-ziptmpd = tempfile.mkdtemp( suffix="-pyenv" )
-tmpzip = os.path.join( ziptmpd, "robot.zip" )
-
-with zipfile.ZipFile(tmpzip, 'w', zipfile.ZIP_DEFLATED) as zip:
-    for folderName, subfolders, filenames in os.walk(tmpd):
-        relativeFolder = os.path.relpath(folderName, tmpd)
-        for filename in filenames:
-            zip.write(os.path.join(folderName, filename), os.path.join(relativeFolder, filename))
-
-
-if os.path.exists( OUTPUT_ZIP ):
-    os.unlink( OUTPUT_ZIP )
-
-shutil.move( tmpzip, OUTPUT_ZIP )
-
-def ensureWritable(function, path, exc_info):
-    os.chmod(path, stat.S_IWRITE)
-    function(path)
-
-shutil.rmtree( ziptmpd, onerror = ensureWritable)
-shutil.rmtree( tmpd, onerror = ensureWritable)
+#!/usr/bin/env python3
+"""
+Utility to create usercode bundles (robot.zip) file compatible with SR kit.
+
+This version of the utility is intended for use with post SR2022 kit, which
+uses the astoria robot management system and the associated usercode format.
+"""
+import argparse
+import os
+import sys
+from fnmatch import fnmatch
+from pathlib import Path
+from typing import Dict, Union
+from zipfile import ZIP_DEFLATED, ZipFile
+
+import toml
+
+ENTRYPOINT = "robot.py"
+IGNORE_PATTERNS = [
+    ".git/*",
+    ".gitignore",
+    "*.pyc",
+    "bundle.toml",  # We don't want to try and include this twice!
+    "robot.zip",
+]
+
+
+class UnableToCreateUsercodeBundleException(Exception):
+    """Unable to create usercode bundle."""
+
+
+def is_file_ignored(filename: str) -> bool:
+    """
+    Check if a file should be ignored and excluded from the zip.
+
+    This is defined using the patterns in IGNORE_PATTERNS.
+
+    :param filename: The filename to check.
+    :return: boolean indicating if the file should be ignored.
+    """
+    return any(fnmatch(filename, pattern) for pattern in IGNORE_PATTERNS)
+
+
+def get_bundle_dict(
+    kit_version: str,
+    wifi_ssid: str,
+    wifi_psk: str,
+    wifi_region: str,
+) -> Dict[str, Dict[str, Union[str, bool]]]:
+    """
+    Get a dictionary of the data for bundle.toml.
+
+    This function should be updated if the bundle format changes.
+
+    :param kit_version: The kit version the bundle is targeted for.
+    :param wifi_ssid: The SSID for the WiFi access point.
+    :param wifi_psk: The PSK for the WiFi access point.
+    :param wifi_region: Region that the WiFi hotspot will be used in.
+    """
+    return {
+        "bundle": {
+            "version": "2.0.0",
+        },
+        "kit": {
+            "name": "Student Robotics",
+            "version": kit_version,
+        },
```

Comment:
What is the purpose of this key?
As I understand this script (unless some part of it is yet to be added?) there is nothing actually here which is going to include kit libraries etc.
It therefore seems a bit odd to be recording this value -- it might as well be arbitrary.

Reply:
See https://srobo.github.io/astoria/design/code_format.html#kit-version

Reply:
Thanks, that tells me a bit more about what the version number might be, but not what it's doing here.

There are two ways around to look at the compatibility of the built bundle:
- the version of the bundle spec
- the version of the kit software that the bundle expects to be used with

The former of these is present in the `"bundle"` key and seems good -- it tells consumers how to read the information contained within the bundle. As it's defined by the tool used to create the bundle it being hardcoded here is fine.

The latter appears to be what this `"kit"` key is, however it doesn't seem to actually achieve anything. There's nothing here which actually checks that the user-code is going to be compatible with the kit version. There's a comment in the code about pulling the version number from an API, though it's not really clear what that means.
For example, is that just pulling the latest version number from somewhere? If so, what purpose does that actually serve?
If the caller of the script were forced to provide this then that might make sense -- they could vouch for the compatibility of the code they're bundling with the kit version. However that's not how this script is configured (nor where it seems to be heading).

If this version number is intended to be a way for the kit software to detect that the user code isn't compatible, there doesn't seem to be any way for it to actually achieve that purpose. This is the part I'm not understanding/assuming there's something missing here.

I suppose another reading of the docs you linked to is that this is a way for the kit code to know that it itself is out of date and to indicate that to the user. If _that_ is the intended use case then this really doesn't feel like it's the right way to achieve it. At the very least I would encourage that the key be called something like "available updates" or something, so that it's clear that it's _data provided within the file_ rather than a _specification of the file_.

The docs however mention that the version number might sometimes be cause for the code not to run, which seems to indicate that this isn't the case and appears to favour the previous explanation.